File: native/GC/ios/Classes/ZohoGcSdkPlugin.swift
import Flutter
import UIKit
import ZohoGCSDK

public class ZohoGcSdkPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "zoho_gc_sdk", binaryMessenger: registrar.messenger())
    let instance = ZohoGcSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }


  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let arguments: [String: Any]? = call.arguments as? Dictionary<String, Any>
    switch ZDGCAPI(rawValue:call.method) {
     case .showFlow:
        showFlow(arguments: arguments)
     case .setSessionVariable:
        setSessionVariable(arguments:arguments)
     case .updateSessionVariable:
        updateSessionVariable(arguments: arguments)
     case .enableLog:
           enableLog(arguments: arguments)
     case .clearData:
            clearData(arguments: arguments)
     case .setTheme:
            setTheme(arguments:arguments)
    case .updateLanguageIcon:
        updateLanguageIcon(arguments: arguments)
    case .hideEndChatPopupWindow:
        hideEndChatPopUp(arguments: arguments)
    case .hideLanguageMismatchPopupWindow:
        hideLanguageMismatchPopupWindow(arguments: arguments)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
    
    public func updateLanguageIcon(arguments: [String: Any]?){
        guard let visibility = arguments?["isVisible"] as? Bool else {return}
        ZohoGC.updateLanguageIcon(isEnable: visibility)
    }

   public func showFlow(arguments: [String: Any]?){
      guard let orgId = arguments?["orgId"] as? String,
              let flowId = arguments?["flowId"] as? String,
              let domain = arguments?["domain"] as? String,
              let preferredLanguage = arguments?["preferredLanguage"] as? String
        else { return }
       ZohoGC.showFlow(orgId: orgId, flowId: flowId, domain: domain, preferredLanguage: preferredLanguage)
   }

   public func setSessionVariable(arguments: [String: Any]?){
      guard let botId = arguments?["botId"] as? String, let arguments = arguments?["sessionVariables"] as? [[String:Any]] else { return }
      ZohoGC.setSessionVariables(botId: botId, sessionVariables: arguments)
   }

    public func updateSessionVariable(arguments: [String: Any]?){
        guard let botId = arguments?["botId"] as? String, let arguments = arguments?["sessionVariables"] as? [[String:Any]] else { return }
        ZohoGC.updateSessionVariables(botId: botId, sessionVariables: arguments)
    }

     public func enableLog(arguments: [String: Any]?){
            guard let isLogEnabled = arguments?["isLogEnabled"] as? Bool
                 else { return }

            ZohoGC.enableLog(isLogEnabled: isLogEnabled)
     }

      public func clearData(arguments: [String: Any]?){
           guard let botId = arguments?["botId"] as? String
                else { return }

           ZohoGC.clearData(botId: botId)
      }

   public func setTheme(arguments: [String: Any]?){
         guard let themeType = arguments?["type"] as? Int,
                let type = ZDThemeType(rawValue: themeType)
              else { return }
         ZDThemeManager.setTheme(type:type)
      }

    public func hideEndChatPopUp(arguments: [String: Any]?) {
        guard let isHide = arguments?["isHide"] as? Bool else {return}
        ZohoGC.hideEndChatPopupWindow(isHide: isHide)
    }

    public func hideLanguageMismatchPopupWindow(arguments: [String: Any]?) {
        guard let isHide = arguments?["isHide"] as? Bool else {return}
        ZohoGC.hideLanguageMismatchPopupWindow(isHide: isHide)
    }

   private enum ZDGCAPI: String {
          case showFlow
          case setSessionVariable
          case updateSessionVariable
          case enableLog
          case clearData
          case setTheme
          case updateLanguageIcon
          case hideEndChatPopupWindow
          case hideLanguageMismatchPopupWindow
    }

}

File: native/GC/android/src/main/kotlin/com/zoho/gc/zoho_gc_sdk/ZohoGcSdkPlugin.kt
package com.zoho.gc.zoho_gc_sdk

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import android.app.Activity
import com.zoho.gc.ZohoGC
import com.zoho.gc.gc_base.ZDChatCallback
import com.zoho.gc.gc_base.ZConfigUtil
import com.zoho.gc.gc_base.ZDThemeType
import com.google.gson.Gson
import android.graphics.Color
import android.util.Log
import java.util.Locale


/** ZohoGcSdkPlugin */
class ZohoGcSdkPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    protected var activity: Activity? = null

    companion object {
        private var darkThemeBuilder: ZDTheme? = null
        private var lightThemeBuilder: ZDTheme? = null
        fun setThemeBuilder(theme: ZDTheme) {
            if (theme.isDarkMode) darkThemeBuilder = theme else lightThemeBuilder = theme
        }
        fun hideLocationSearch(value : Boolean) {
            ZConfigUtil.hideLocationSearch = value
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "zoho_gc_sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val arguments = call.arguments as? HashMap<String, Any>?
        when (call.method) {
            "enableLog" -> enableLog(arguments)
            "showFlow" -> showFlow(arguments)
            "setSessionVariable" -> setSessionVariable(arguments)
            "updateSessionVariable" -> updateSessionVariable(arguments)
            "clearData" -> clearData(arguments)
            "setTheme" -> setTheme(arguments)
            "setLocale" -> setLocale(arguments)
        }
    }


    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }


    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    fun showFlow(arguments: HashMap<String, Any>?) {
        val orgId = arguments?.get("orgId") as? String ?: ""
        val botId = arguments?.get("flowId") as? String ?: ""
        val domain = arguments?.get("domain") as? String ?: ""
        val preferredLanguage = arguments?.get("preferredLanguage") as? String ?: ""
        activity?.let {
            ZohoGC.showFlow(it, orgId, botId, domain, preferredLanguage)
        }
    }

    fun setSessionVariable(arguments: HashMap<String, Any>?) {
        val botId = arguments?.get("botId") as? String ?: ""
        val sessionVariables =
            arguments?.get("sessionVariables") as? ArrayList<HashMap<String, Any>> ?: arrayListOf()
        ZohoGC.setSessionVariables(botId, sessionVariables)
    }

    fun updateSessionVariable(arguments: HashMap<String, Any>?) {
        val orgId = arguments?.get("orgId") as? String ?: ""
        val botId = arguments?.get("botId") as? String ?: ""
        val domain = arguments?.get("domain") as? String ?: ""
        val sessionVariables =
            arguments?.get("sessionVariables") as? ArrayList<HashMap<String, Any>> ?: arrayListOf()
        activity?.let {
            ZohoGC.updateSessionVariables(it, orgId, botId, domain, sessionVariables)
        }
    }

    fun clearData(arguments: HashMap<String, Any>?) {
        val botId = arguments?.get("botId") as? String ?: ""
        activity?.let {
            ZohoGC.clearData(it, botId, object : ZDChatCallback.ZDClearDataCallback {
                override fun onFailed(exception: Exception) {}
                override fun onSuccess() {}
            })
        }
    }

    fun enableLog(arguments: HashMap<String, Any>?) {
        val isLogEnabled = arguments?.get("isLogEnabled") as? Boolean ?: false
        ZohoGC.enableLog(isLogEnabled)
    }

    fun setTheme(arguments: HashMap<String, Any>?) {
        val gson = Gson()
        val themeType = when (arguments?.get("type") as Int) {
            1 -> ZDThemeType.DARK
            0 -> ZDThemeType.LIGHT
            else -> ZDThemeType.SYSTEM
        }
        ZConfigUtil.setThemeType(themeType);
        lightThemeBuilder?.let { lightTheme ->
            ZConfigUtil.setThemeBuilder(
                gson.fromJson(
                    gson.toJson(lightTheme),
                    com.zoho.gc.gc_base.ZDTheme::class.java
                )
            )
        }
        darkThemeBuilder?.let { darkTheme ->
            ZConfigUtil.setThemeBuilder(
                gson.fromJson(
                    gson.toJson(darkTheme),
                    com.zoho.gc.gc_base.ZDTheme::class.java
                )
            )
        }
    }

    fun setLocale(arguments: HashMap<String, Any>?) {
        val languageCode = arguments?.get("languageCode") as? String ?: Locale.getDefault().language
        val countryCode = arguments?.get("countryCode") as? String ?: Locale.getDefault().country
        val locale = Locale(languageCode, countryCode)
        ZConfigUtil.locale = locale
    }

}
